# Cours 6: Machine learning

---

## 5 - 1 Théorie

*Cette section est directement issue de mes propres notes. Je traduirai quand je peux, mais quasiment toutes les ressources sur le machine learning sont en anglais, prenez l'habitude.*

**Machine learning**: subfield of computer science that gives computers the ability to learn without being explicitly programmed
- A computer program is said to learn from experience E with respect to some class of tasks T and performance measure P if its performance at tasks in T, as measured by P, improves with experience E.

### Two categories of ML

**Supervised**: labeled data => determine desired outcome from specific input

- Regression: output is a continuous number => response is continuous
    - Xylose metabolism expression levels in function of availability of xylose in the environment
- Classification: output belongs to discrete, finite set => response is categorical
    - Image recognition
- Outline
    - Clean data and put it into machine-readable form
    - Develop machine learning model
    - Train model on training data
        - Model learns to map attributes of data on outcomes
    - Test model on separate set of data
    - Make predictions on new data


**Unsupervised**: unlabeled data => determine structure and relations between the training examples
- Clustering algorithms
    - Finding similar instances or groups of instances
- Non-clustering algorithms


### Machine learning algorithms

- Different types
    - Use different techniques to do the learning
    - Each algorithm is better suited to a specific type of task
- Machine learning algorithms try to make generalizations based on data
    - If an algorithm is highly specific to one type of data, it cannot be called machine learning
    - A good model can generalize what it learns from training data to new, untrained data
    - A common problem in machine learning is overfitting
        - Overfitting: model is excessively tuned to training data, has low prediction performance on new, untrained data
- Examples
    - Naive Bayes
        - Supervised learning algorithm
        - Probabilistic model
            - => predict response by modeling problem with a probability distribution
        - Based on applying Bayes' theorem with strong independence assumptions between the features
            - => features are considered independent
            - Output => prediction + degree of certainty
    - Support vector machine/SVM
        - Supervised learning algorithm
        - Builds set of hyperplanes that separate instances of different classes
            - Hyperplane = maximal distance between the closest data points in 2 classes
                - => largest separation between nearest instances from different classes
    - K-means
        - Unsupervised learning algorithm
        - Partitions data into k different clusters based on data features
            - Clusters gather data points so that each data point in cluster has minimum distance to mean of cluster
            - k => fixed by user

## 5 - 2 Scikit-learn

### 5 - 2 - 1 Classification

Nous allons prendre un dataset très connu par les statisticiens, l'Iris dataset.

In [1]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.cross_validation import train_test_split
from sklearn.datasets import load_iris

dataset = load_iris()

print(dataset.keys())
print("\n")
print(dataset["feature_names"])
print("\n")
print(dataset["target_names"])
print("\n")
print(dataset["DESCR"])

dict_keys(['data', 'target_names', 'target', 'feature_names', 'DESCR'])


['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


['setosa' 'versicolor' 'virginica']


Iris Plants Database

Notes
-----
Data Set Characteristics:
    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9  

In [2]:
X = dataset.data
Y = dataset.target

print(X.shape)

print(Y.shape)

(150, 4)
(150,)


Nous avons 150 observations de 4 features (longueur et largeur des sépales et pétales), classifiées dans 3 catégories (setosa, versicolor, virginica).

Nous allons créer notre training set (pour entraîner les modèles) et notre testing set (pour faire les prédictions).

In [3]:
x_train, x_test, y_train, y_test = train_test_split(X, Y, train_size=0.75)

print(x_train.shape, y_train.shape, x_test.shape, y_test.shape)

print(x_train[:5])

print(y_train[:5])

(112, 4) (112,) (38, 4) (38,)
[[ 6.4  2.9  4.3  1.3]
 [ 6.4  2.8  5.6  2.1]
 [ 7.4  2.8  6.1  1.9]
 [ 4.8  3.1  1.6  0.2]
 [ 5.6  2.5  3.9  1.1]]
[1 2 2 0 1]


In [4]:
model = DecisionTreeClassifier()

model.fit(x_train, y_train)

model.score(x_test, y_test)

0.92105263157894735

Le modèle Decision Tree a une précision de 92%.

In [27]:
model = RandomForestClassifier()

model.fit(x_train, y_train)

model.score(x_test, y_test)

0.94736842105263153

Le modèle Random Forest (un ensemble de Decision Trees) a une précision de 95%. Ces 2 modèles sont les plus utilisés pour la classification.

### 5 - 2 - 2 Regression

## 5 - 3 Tensorflow + Keras